File: java-sdk/radar-schemas-core/src/main/java/org/radarbase/schema/SchemaCatalogue.java
package org.radarbase.schema;

import kotlin.Pair;
import org.apache.avro.Schema;
import org.apache.avro.Schema.Parser;
import org.apache.avro.generic.GenericRecord;
import org.radarbase.config.AvroTopicConfig;
import org.radarbase.schema.validation.SchemaValidator;
import org.radarbase.schema.validation.rules.SchemaMetadata;
import org.radarbase.topic.AvroTopic;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

import static java.util.function.Function.identity;
import static java.util.function.Predicate.not;

public class SchemaCatalogue {
    private final Path root;
    private final Map<String, SchemaMetadata> schemas;
    private final List<SchemaMetadata> unmappedFiles;


    public SchemaCatalogue(Path root) throws IOException {
        this(root, null);
    }

    public SchemaCatalogue(Path root, Scope scope) throws IOException {
        this.root = root.resolve("commons");
        Map<String, SchemaMetadata> schemaTemp = new HashMap<>();
        List<SchemaMetadata> unmappedTemp = new ArrayList<>();

        if (scope != null) {
            loadSchemas(schemaTemp, unmappedTemp, scope);
        } else {
            for (Scope useScope : Scope.values()) {
                loadSchemas(schemaTemp, unmappedTemp, useScope);
            }
        }

        schemas = Collections.unmodifiableMap(schemaTemp);
        unmappedFiles = Collections.unmodifiableList(unmappedTemp);
    }

    /**
     * Returns an avro topic with the schemas from this catalogue.
     * @param config avro topic configuration
     * @return AvroTopic with
     * @throws NoSuchElementException if the key or value schema do not exist in this catalogue.
     * @throws NullPointerException if the key or value schema configurations are null
     * @throws IllegalArgumentException if the topic configuration is null
     */
    public AvroTopic<GenericRecord, GenericRecord> getGenericAvroTopic(AvroTopicConfig config) {
        Pair<SchemaMetadata, SchemaMetadata> schemaMetadata = getSchemaMetadata(config);
        return new AvroTopic<>(config.getTopic(),
                schemaMetadata.component1().getSchema(), schemaMetadata.component2().getSchema(),
                GenericRecord.class, GenericRecord.class);
    }

    public Map<String, SchemaMetadata> getSchemas() {
        return schemas;
    }

    public List<SchemaMetadata> getUnmappedAvroFiles() {
        return unmappedFiles;
    }

    private void loadSchemas(
            Map<String, SchemaMetadata> schemas,
            List<SchemaMetadata> unmappedFiles,
            Scope scope) throws IOException {

        Path walkRoot = scope.getPath(root);
        if (walkRoot == null) {
            return;
        }

        List<Path> avroFiles = Files.walk(walkRoot)
                .filter(p -> Files.isRegularFile(p) && SchemaValidator.isAvscFile(p))
                .collect(Collectors.toList());

        int prevSize = -1;

        // Recursively parse all schemas.
        // If the parsed schema size does not change anymore, the final schemas cannot be parsed
        // at all.
        while (prevSize != schemas.size()) {
            prevSize = schemas.size();
            Map<String, Schema> useTypes = schemas.entrySet().stream()
                    .collect(Collectors.toMap(Entry::getKey, e -> e.getValue().getSchema()));
            Set<Path> ignoreFiles = schemas.values().stream()
                    .map(SchemaMetadata::getPath)
                    .collect(Collectors.toSet());

            schemas.putAll(avroFiles.stream()
                    .filter(not(ignoreFiles::contains))
                    .map(p -> {
                        @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
                        Parser parser = new Parser();
                        parser.addTypes(useTypes);
                        try {
                            return new SchemaMetadata(parser.parse(p.toFile()), scope, p);
                        } catch (Exception ex) {
                            return null;
                        }
                    })
                    .filter(Objects::nonNull)
                    .collect(Collectors.toMap(
                            m -> m.getSchema().getFullName(),
                            identity(),
                            (v1, v2) -> {
                                if (v1.equals(v2)) {
                                    return v1;
                                } else {
                                    throw new IllegalStateException("Duplicate enum: " + v1);
                                }
                            })));
        }

        Set<Path> mappedPaths = schemas.values().stream()
                .map(SchemaMetadata::getPath)
                .collect(Collectors.toSet());

        unmappedFiles.addAll(avroFiles.stream()
                .filter(p -> !mappedPaths.contains(p))
                .map(p -> new SchemaMetadata(null, scope, p))
                .collect(Collectors.toList()));
    }

    /**
     * Returns an avro topic with the schemas from this catalogue.
     * @param config avro topic configuration
     * @return AvroTopic with
     * @throws NoSuchElementException if the key or value schema do not exist in this catalogue.
     * @throws NullPointerException if the key or value schema configurations are null
     * @throws IllegalArgumentException if the topic configuration is null
     */
    public Pair<SchemaMetadata, SchemaMetadata> getSchemaMetadata(AvroTopicConfig config) {
        SchemaMetadata parsedKeySchema = schemas.get(config.getKeySchema());
        if (parsedKeySchema == null) {
            throw new NoSuchElementException("Key schema " + config.getKeySchema()
                    + " for topic " + config.getTopic() + " not found.");
        }

        SchemaMetadata parsedValueSchema = schemas.get(config.getValueSchema());
        if (parsedValueSchema == null) {
            throw new NoSuchElementException("Value schema " + config.getValueSchema()
                    + " for topic " + config.getTopic() + " not found.");
        }

        return new Pair<>(parsedKeySchema, parsedValueSchema);
    }
}

File: java-sdk/radar-schemas-core/src/main/java/org/radarbase/schema/SchemaCatalogue.kt
package org.radarbase.schema

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.apache.avro.Schema
import org.apache.avro.generic.GenericRecord
import org.radarbase.config.AvroTopicConfig
import org.radarbase.kotlin.coroutines.forkJoin
import org.radarbase.schema.specification.config.SchemaConfig
import org.radarbase.schema.util.SchemaUtils.listRecursive
import org.radarbase.schema.validation.SchemaValidator.Companion.isAvscFile
import org.radarbase.schema.validation.rules.FailedSchemaMetadata
import org.radarbase.schema.validation.rules.SchemaMetadata
import org.radarbase.topic.AvroTopic
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.Path
import java.nio.file.PathMatcher
import java.util.*
import kotlin.io.path.exists
import kotlin.io.path.inputStream

class SchemaCatalogue @JvmOverloads constructor(
    private val schemaRoot: Path,
    config: SchemaConfig,
    scope: Scope? = null,
) {
    val schemas: Map<String, SchemaMetadata>
    val unmappedAvroFiles: List<FailedSchemaMetadata>

    init {
        val schemaTemp = HashMap<String, SchemaMetadata>()
        val unmappedTemp = mutableListOf<FailedSchemaMetadata>()
        val matcher = config.pathMatcher(schemaRoot)
        runBlocking {
            if (scope != null) {
                loadSchemas(schemaTemp, unmappedTemp, scope, matcher, config)
            } else {
                for (useScope in Scope.entries) {
                    loadSchemas(schemaTemp, unmappedTemp, useScope, matcher, config)
                }
            }
        }
        schemas = schemaTemp.toMap()
        unmappedAvroFiles = unmappedTemp.toList()
    }

    /**
     * Returns an avro topic with the schemas from this catalogue.
     * @param config avro topic configuration
     * @return AvroTopic with
     * @throws NoSuchElementException if the key or value schema do not exist in this catalogue.
     * @throws NullPointerException if the key or value schema configurations are null
     * @throws IllegalArgumentException if the topic configuration is null
     */
    fun getGenericAvroTopic(config: AvroTopicConfig): AvroTopic<GenericRecord, GenericRecord> {
        val (keySchema, valueSchema) = getSchemaMetadata(config)
        return AvroTopic(
            requireNotNull(config.topic) { "Missing Avro topic in configuration" },
            requireNotNull(keySchema.schema) { "Missing Avro key schema" },
            requireNotNull(valueSchema.schema) { "Missing Avro value schema" },
            GenericRecord::class.java,
            GenericRecord::class.java,
        )
    }

    @Throws(IOException::class)
    private suspend fun loadSchemas(
        schemas: MutableMap<String, SchemaMetadata>,
        unmappedFiles: MutableList<FailedSchemaMetadata>,
        scope: Scope,
        matcher: PathMatcher,
        config: SchemaConfig,
    ) {
        val walkRoot = schemaRoot.resolve(scope.lower)
        val avroFiles = buildMap<Path, String> {
            if (walkRoot.exists()) {
                walkRoot
                    .listRecursive { matcher.matches(it) && it.isAvscFile() }
                    .forkJoin(Dispatchers.IO) { p ->
                        p.inputStream().reader().use {
                            p to it.readText()
                        }
                    }
                    .toMap(this@buildMap)
            }
            config.schemas(scope).forEach { (key, value) ->
                put(walkRoot.resolve(key), value)
            }
        }

        var prevSize = -1

        // Recursively parse all schemas.
        // If the parsed schema size does not change anymore, the final schemas cannot be parsed
        // at all.
        while (prevSize != schemas.size) {
            prevSize = schemas.size
            val useTypes = schemas.mapValues { (_, v) -> v.schema }
            val ignoreFiles = schemas.values.mapTo(HashSet()) { it.path }

            schemas.putParsedSchemas(avroFiles, ignoreFiles, useTypes, scope)
        }
        val mappedPaths = schemas.values.mapTo(HashSet()) { it.path }

        avroFiles.keys.asSequence()
            .filter { it !in mappedPaths }
            .distinct()
            .mapTo(unmappedFiles) { p -> FailedSchemaMetadata(scope, p) }
    }

    private suspend fun MutableMap<String, SchemaMetadata>.putParsedSchemas(
        customSchemas: Map<Path, String>,
        ignoreFiles: Set<Path>,
        useTypes: Map<String, Schema>,
        scope: Scope,
    ) = customSchemas
        .filter { (p, _) -> p !in ignoreFiles }
        .entries
        .forkJoin { (p, schema) ->
            val parser = Schema.Parser()
            parser.addTypes(useTypes)
            withContext(Dispatchers.IO) {
                try {
                    val parsedSchema = parser.parse(schema)
                    parsedSchema.fullName to SchemaMetadata(parsedSchema, scope, p)
                } catch (ex: Exception) {
                    logger.debug("Cannot parse schema {}: {}", p, ex.toString())
                    null
                }
            }
        }
        .filterNotNull()
        .toMap(this@putParsedSchemas)

    /**
     * Returns an avro topic with the schemas from this catalogue.
     * @param config avro topic configuration
     * @return AvroTopic with
     * @throws NoSuchElementException if the key or value schema do not exist in this catalogue.
     * @throws NullPointerException if the key or value schema configurations are null
     * @throws IllegalArgumentException if the topic configuration is null
     */
    fun getSchemaMetadata(config: AvroTopicConfig): Pair<SchemaMetadata, SchemaMetadata> {
        val parsedKeySchema = schemas[config.keySchema]
            ?: throw NoSuchElementException(
                "Key schema " + config.keySchema +
                    " for topic " + config.topic + " not found.",
            )
        val parsedValueSchema = schemas[config.valueSchema]
            ?: throw NoSuchElementException(
                "Value schema " + config.valueSchema +
                    " for topic " + config.topic + " not found.",
            )
        return Pair(parsedKeySchema, parsedValueSchema)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(SchemaCatalogue::class.java)
    }
}
